README.md
Copyright (c) 2009-2011 by Appcelerator, Inc. All Rights Reserved.
Licensed under the terms of the Apache Public License
Please see the LICENSE included with this distribution for details.

# nodejs-desktop-prototype README #

nodejs-desktop-prototype is a prototype application that demonstates packaging and running a local Node.js server with Titanium Desktop applications.  [Appcelerator Titanium Desktop](http://www.appcelerator.com/products/titanium-desktop-application-development/) is a free and open source application development platform that lets you create native desktop application experiences using existing web skills like Javascript, HTML, 
CSS, Python, Ruby, and PHP.

![nodejs-desktop-prototype screenshot](http://developer.appcelerator.com.s3.amazonaws.com/documentation-examples/nodejs-desktop-prototype.png)


### Additional Notes
nodejs-desktop-prototype makes use of:

* [node](http://nodejs.org/) and the [node-github module](https://github.com/ajaxorg/node-github)
* [EJS](http://embeddedjs.com/)
* [jQuery](http://jquery.com/) and [jQuery UI](http://jqueryui.com/)

Special thanks to goes out to these great projects.

### Background
* This is a prototype (POC if you will)
* Windows does not work yet due to limitations with Node.js on Windows
* This has been tested with OSX (Snow Leopard) & Ubuntu 10.10.


CHANGELOG.txt
Version 1.0 (06/30/2011):
* Initial release

Resources/index.js
var fetchRepoList = function() {
	$.get('http://127.0.0.1:1338/get_repos', function(data) {
		var targetNode = $('#repositories'),
			jsonData   = $.parseJSON(data);
    	var repoList = new EJS({url: 'templates/repos.ejs'}).render(jsonData);
    	targetNode.html(repoList);
    	
    	$("#repolist").delegate(".repolist-link", "click", function(e){
	    	e.preventDefault();
	    	
	    	if ($(this).hasClass('repo') === true) {
	    		Titanium.Platform.openURL($(this).attr('href'));
	    	} else if ($(this).hasClass('watchercount') === true) {
	    		fetchWatchersList($(this).attr('rel'));
	    	}
	    });
	});
};

var fetchFollowersList = function() {
	$.get('http://127.0.0.1:1338/get_followers', function(data) {
		var jsonData   = $.parseJSON(data);
	    
        html = new EJS({url: 'templates/userlist.ejs'}).render(jsonData);
        generalContentOverlay.html(html);
    	generalContentOverlay.dialog('option', 'title', 'Github Followers');
    	generalContentOverlay.dialog('open');
	});
};

var fetchWatchersList = function(reponame) {
	$.get('http://127.0.0.1:1338/get_repo_watchers/' + reponame, function(data) {
		var jsonData   = $.parseJSON(data);
	    
        if (jsonData) {
        	html = new EJS({url: 'templates/userlist.ejs'}).render(jsonData);
        	generalContentOverlay.html(html);
	    	generalContentOverlay.dialog('option', 'title', 'Repo Watchers');
	    	generalContentOverlay.dialog('open');
        }
	});
};

var fetchProfile = function(username) {
	$.get('http://127.0.0.1:1338/get_profile/' + username, function(data) {
		var html,
			jsonData = $.parseJSON(data);
	    
        if (jsonData) {
        	html = new EJS({url: 'templates/profile.ejs'}).render(jsonData);
        	generalContentOverlay.html(html);
    		generalContentOverlay.dialog('option', 'title', 'User Profile');
        }
	});
};

var generalContentOverlay;
$(window).load(function() {
	generalContentOverlay = $("#general-content-overlay");
	
	generalContentOverlay.dialog({
		height: 400,
		width: 500,
		modal: true,
		resizable: false,
		draggable: false,
		autoOpen: false
	});
	
	$(window).resize(function() {
    	generalContentOverlay.dialog("option", "position", "center");
	});
	$(window).scroll(function() {
	    generalContentOverlay.dialog("option", "position", "center");
	});
	
	$('#fetch-followers').click(function(e) {
		e.preventDefault();
		fetchFollowersList();
	});
    
    $("#general-content-overlay").delegate("a.user", "click", function(e){
    	e.preventDefault();
		fetchProfile(e.target.rel);
    });
    
    $("#general-content-overlay").delegate("a.blog-link", "click", function(e){
    	e.preventDefault();
		Titanium.Platform.openURL($(this).attr('href'));
    });
});

Resources/nodejs/nodejs.py
import os, subprocess, time

class TiNodeJS:
    def __init__(self):
        self.platform = Titanium.getPlatform()
        if self.platform == 'win':
            alert('Sorry the Windows platform is not supported yet!');
            return
        
        self.nodepid = None
        self.nodeproc = None
        self.sep = Titanium.Filesystem.getSeparator()
        self.noderoot = Titanium.App.getHome() + self.sep + 'Resources' + self.sep + 'nodejs'
        self.nodebinary = self.noderoot + self.sep + 'builds' + self.sep + self.platform + self.sep + 'node' + self.sep + 'bin' + self.sep + 'node'
        self.pathtoappserver = Titanium.App.getHome() + self.sep + 'Resources' + self.sep + 'appserver.js'
    
    def startup(self, callback):
        print '@@@ Doing tinode startup!'
        
        #The first time through we'll need to set the execution bit
        if not os.access(self.nodebinary, os.X_OK):
            os.chmod(self.nodebinary, 744)
        
        self.nodeproc = subprocess.Popen(['nohup', self.nodebinary,  self.pathtoappserver])
        self.nodepid = self.nodeproc.pid
        print('Node started pid = %d' % self.nodepid)

        time.sleep(1) #Give server a moment to start listening (@TODO: actually verify it is listening)
        callback()

    def shutdown(self):
        print '@@@ Doing tinode shutdown!'
        
        if self.nodepid is not None:
            print('Killing Node pid = %d' % self.nodepid)
            
            #self.nodeproc.terminate()
            #self.nodeproc.wait()
            os.kill(self.nodepid, 6)
            os.wait()

Resources/nodejs/builds/linux/node/include/node/config.h
/* Configuration header created by Waf - do not edit */
#ifndef _CONFIG_H_WAF
#define _CONFIG_H_WAF

#define HAVE_OPENSSL 1
#define HAVE_PTHREAD_CREATE 1
#define HAVE_PTHREAD_ATFORK 1
#define HAVE_FUTIMES 1
#define HAVE_READAHEAD 1
#define HAVE_FDATASYNC 1
#define HAVE_PREADWRITE 1
#define HAVE_SENDFILE 1
#define HAVE_SYNC_FILE_RANGE 1
#define HAVE_SYS_INOTIFY_H 1
#define HAVE_INOTIFY_INIT 1
#define HAVE_SYS_EPOLL_H 1
#define HAVE_EPOLL_CTL 1
/* #undef HAVE_PORT_H */
#define HAVE_POLL_H 1
#define HAVE_POLL 1
/* #undef HAVE_SYS_EVENT_H */
#define HAVE_SYS_QUEUE_H 1
/* #undef HAVE_KQUEUE */
#define HAVE_SYS_SELECT_H 1
#define HAVE_SELECT 1
#define HAVE_SYS_EVENTFD_H 1
#define HAVE_EVENTFD 1
#define HAVE_CLOCK_SYSCALL 1
#define HAVE_CLOCK_GETTIME 1
#define HAVE_NANOSLEEP 1
#define HAVE_CEIL 1
#define HAVE_CONFIG_H 1
#endif /* _CONFIG_H_WAF */


Resources/nodejs/builds/linux/node/include/node/node_config.h
// Copyright Joyent, Inc. and other Node contributors.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.

#ifndef NODE_CONFIG_H
#define NODE_CONFIG_H

#define NODE_CFLAGS "-rdynamic -D_GNU_SOURCE -DHAVE_CONFIG_H=1 -pthread -m32 -g -O3 -DHAVE_OPENSSL=1 -DEV_FORK_ENABLE=0 -DEV_EMBED_ENABLE=0 -DEV_MULTIPLICITY=0 -DX_STACKSIZE=65536 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DEV_MULTIPLICITY=0 -DHAVE_FDATASYNC=1 -DPLATFORM=\"linux\" -D__POSIX__=1 -Wno-unused-parameter -D_FORTIFY_SOURCE=2 -I/home/chad/local/node/include/node"
#define NODE_PREFIX "/home/chad/local/node"

#endif /* NODE_CONFIG_H */


Resources/appserver.js
var server,
	http      = require('http'),
	sys       = require("sys"),
	GitHubApi = require("github").GitHubApi,
	github    = new GitHubApi(true),
	githubUserOfConcern = 'appcelerator';
	
var ohNo = function(req, resp) {
	resp.writeHead(500, {"Content-Type": "text/plain"});  
    resp.write(err + "\n");  
    resp.close();
    return;
};
	
var getFollowers = function(req, resp) {
	github.getUserApi().getFollowers(githubUserOfConcern, function(err, followers) {
	    if(err) { ohNo(); }
	    
	    if (followers) {
	    	resp.write(JSON.stringify({users: followers}));
	    } else {
	    	resp.write(JSON.stringify({users: []}));
	    }
	    
	    resp.end();
	});
};

var getRepos = function(req, resp) {
	github.getRepoApi().getUserRepos(githubUserOfConcern, function(err, repositories) {
	    if(err) { ohNo(); }

	    if (repositories) {
	    	resp.write(JSON.stringify({repos: repositories}));
	    } else {
	    	resp.write(JSON.stringify({repos: []}));
	    }
	    
	    resp.end();
	});
};

var getRepoWatchers = function(req, resp) {
	var urlParts = req.url.split('/');
	if (urlParts && urlParts.length === 3) {
		//Valid length, so pick out the repo name
		var repo = urlParts[2];
		
		github.getRepoApi().getRepoWatchers(githubUserOfConcern, repo, function(err, watchers) {
		    if(err) { ohNo(); }

		    if (watchers) {
		    	resp.write(JSON.stringify({users: watchers}));
		    } else {
		    	resp.write(JSON.stringify({users: []}));
		    }
		    
		    resp.end();
		});
	} else {
		resp.writeHead(404, {"Content-Type": "text/plain"});  
        resp.write("404 Not Found\n");  
        resp.end(); 
        return;
	}
};

var getProfile = function(req, resp) {
	var urlParts = req.url.split('/');
	if (urlParts && urlParts.length === 3) {
		//Valid length, so pick out the username
		var username = urlParts[2];
		
		github.getUserApi().show(username, function(err, details) {
		    if(err) { ohNo(); }
		    
		    resp.write(JSON.stringify(details));
		    resp.end();
		});
	} else {
		resp.writeHead(404, {"Content-Type": "text/plain"});  
        resp.write("404 Not Found\n");  
        resp.end(); 
        return;
	}
};

server = http.createServer(function (req, resp) {
	resp.writeHead(200, {'Content-Type': 'text/plain'});
	
	if (req.url === '/get_followers') {
		getFollowers(req, resp);
	} else if (req.url === '/get_repos') {
		getRepos(req, resp);
	} else if (req.url.indexOf('/get_repo_watchers') === 0) {
		getRepoWatchers(req, resp);
	} else if (req.url.indexOf('/get_profile') === 0) {
		getProfile(req, resp);
	} else {
		resp.writeHead(404, {"Content-Type": "text/plain"});  
        resp.write("404 Not Found\n");  
        resp.end(); 
        return;
	}
}).listen(1338, "127.0.0.1");

console.log('NodeJS server now running at http://127.0.0.1:1338/');